README.md
<div align="center">
	<img src="https://github.com/alumna/liven/raw/master/liven.svg?sanitize=true" alt="liven" width="480" height="270" />
</div>

<div align="center">
	<a href="https://npmjs.org/package/@alumna/liven">
		<img src="https://badgen.now.sh/npm/v/@alumna/liven" alt="version" />
	</a>
	<a href="https://npmjs.org/package/@alumna/liven">
		<img src="https://badgen.net/bundlephobia/min/@alumna/liven" alt="size" />
	</a>
	<a href="https://travis-ci.org/alumna/liven">
		<img src="https://travis-ci.org/alumna/liven.svg?branch=master" alt="travis" />
	</a>
	<a href="https://codecov.io/gh/alumna/liven">
		<img src="https://codecov.io/gh/alumna/liven/branch/master/graph/badge.svg" />
	</a>
	<a href="https://npmjs.org/package/@alumna/liven">
		<img src="https://badgen.now.sh/npm/dm/@alumna/liven" alt="downloads" />
	</a>
</div>

<div align="center">Optimized and lightweight live reload library on 54kB, without dependencies</div>

<br/>

## Features

* Production version with **no dependencies**
* Extremely lightweight - 52kB!
* Alternative to `browser-sync` and `livereload`
* Excellent option to be used inside other libraries
* 100% tested and working on Linux, Mac and Windows!

Additionally, this module is delivered as:

* **ES Module**: [`dist/liven.es.js`](https://unpkg.com/@alumna/liven/dist/liven.es.js)
* **CommonJS**: [`dist/liven.cjs.js`](https://unpkg.com/@alumna/liven/dist/liven.cjs.js)


## Install

```
$ npm install @alumna/liven
```


## Usage

```js
import liven from '@alumna/liven';

const instance = await liven({

	// (OPTIONAL)
	// The current project's directory '.' will be used if `dir` isn't passed
	dir: './public_html/',

	// (OPTIONAL)
	// Port to use on the http server
	// If it isn't available, the first available port between 3000 and 3100 will be used 
	port: 3000,

	// (OPTIONAL)
	// Run as a SPA, mapping non-existent URL's to [base]index.html
	// Default to "false"
	spa: false,

	// (OPTIONAL)
	// Function (as string) to be runned on the browser when receiving a socket signal
	// Defaults to: 'function( data ){ location.reload( true ) };', that simply reloads the page
	// `data` is an object: { path, isDir, isFile, isNew, add_or_update }
	script: 'function( data ){ location.reload( true ) };',

	// (OPTIONAL)
	// Mask paths to a different directory when requested
	alias: {
		'images'    : 'optimized'
	},
	
	// (OPTIONAL)
	// Function to be called on each event, before the refresh signal
	// The function must return "true" to allow the refresh, of "false" to don't allow
	on_event: ({ path, isDir, isFile, isNew, add_or_update }) => {...},
	
});

// You can programatically force a refresh as well
instance.refresh({ path, isDir, isFile, add_or_update, isNew })


// Create a memory-file on the server
// It overwrites the file on the same path, if it exists, on future requests
// -
// "content" can be a string or a buffer
// -
// Path must be a fullpath or initiate without a slash ('/') when relative
// Will generate a refresh unless "on_event" function returns "false" for the path
// -
// It will be instantly available in all modes, including:
// SPA, index.html served on dir URL's, updating 404 files, etc
await instance.memory( path, content )

// Clear a memory-file on the server
// -
// If a file exists on the same path on disk,
// its original content will be served again on future requests
// -
// As on 'memory' feature:
// Path must be a fullpath or initiate without a slash ('/') when relative
// Will generate a refresh unless "on_event" function returns "false" for the path
// -
// The effect will be instantly available in all modes
await instance.clear( path )


// And get the port used
console.log( instance.port ) // 3000
```


CHANGELOG.md
# Liven changelog

## 1.4.0 - 2020-07-22

* Added the feature to clear previously added in-memory files
* Updated devDependencies

## 1.3.1 - 2020-03-12

* Updated devDependencies

## 1.3.0 - 2019-07-19

* New feature: In-memory files

## 1.2.0 - 2019-07-19

* New feature: Alias
* Replace Sirv/Polka with Pulsa

## 1.1.1 - 2019-06-10

* Code quality: removed unused variable from polka.listen()

## 1.1.0 - 2019-06-10

* New fature: reload script is now automatically added/injected 

## 1.0.0 - 2019-06-05

* First release of Liven

test/liven.test.js
import liven from './../src/index';

import * as fs 				from 'fs';
import http 				from 'http';
import { promisify } 		from 'util';

const read	  = promisify( fs.readFile );
const unlink  = promisify( fs.unlink );
const write	  = promisify( fs.writeFile );

const sleep = ms => new Promise( resolve => setTimeout( resolve, ms ) );

describe('Liven tests', () => {

	test('1. Refresh when file change', async ( done ) => {

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'test/01/index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		await sleep( 130 )

		const server  = await liven( { dir: 'test/01' } );

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		// Wait the page to refresh
		await page.waitForNavigation( { waitUntil: 'load' } )

		await expect( page.title() ).resolves.toMatch( 'New title' );

		// Undo the changes
		await write( file, content );

		done();

	});

	test('2. Refresh after "on_event" returning "true"', async ( done ) => {

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'test/02/index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		await sleep( 130 )

		const server  = await liven( { dir: 'test/02', on_event: () => true } );

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		// Wait the page to refresh
		await page.waitForNavigation( { waitUntil: 'load' } )

		await expect( page.title() ).resolves.toMatch( 'New title' );

		// Undo the changes
		await write( file, content );

		done();

	});

	test('3. Do not refresh after "on_event" returning "false"', async ( done ) => {

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'test/03/index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		await sleep( 130 )

		const server  = await liven( { dir: 'test/03', on_event: () => false } );

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Undo the changes
		await write( file, content );

		done();

	});

	test('4. Refresh when file change, with a filter on watcher', async ( done ) => {

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'test/04/index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		await sleep( 130 )

		const server  = await liven( { dir: 'test/04', filter: () => true } );

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		// Wait the page to refresh
		await page.waitForNavigation( { waitUntil: 'load' } )

		await expect( page.title() ).resolves.toMatch( 'New title' );

		// Undo the changes
		await write( file, content );

		done();

	});

	test('5. Refresh when a file is removed', async ( done ) => {

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'test/05/index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		const to_delete = 'test/05/delete_test'
		await write( to_delete, 'delete test', 'utf8' );
		// fs.closeSync( fs.openSync( to_delete, 'w' ) );
		// await write( to_delete, 'delete test', 'utf8' );

		await sleep( 130 )

		const server  = await liven( { dir: 'test/05', filter: ( { path } ) => { return path == 'index.html' ? false : true } } );

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		// Generate a refresh deleting "delete_test"
		await unlink( to_delete );

		// Wait the page to refresh
		await page.waitForNavigation( { waitUntil: 'load' } );

		await expect( page.title() ).resolves.toMatch( 'New title' );

		// Undo the changess
		await write( file, content );
		await write( to_delete, 'delete test', 'utf8' );

		done();

	});

	test('6. Refresh when a file is created', async ( done ) => {

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'test/06/index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		// Delete the file "create_test" if it exists
		const to_create = 'test/06/create_test'
		try {
			await unlink( to_create );
		} catch ( error ) {
		    // continue
		}

		try {
			await unlink( to_create );
		} catch ( error ) {
		    // continue
		}

		await sleep( 130 )

		const server  = await liven( { dir: 'test/06', filter: ( { path } ) => { return path == 'index.html' ? false : true } } );

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		// Generate a refresh deleting "create_test"
		fs.closeSync( fs.openSync( to_create, 'w' ) );
		await write( to_create, 'create test', 'utf8' );

		// Wait the page to refresh
		// await sleep( 130 )
		await page.waitForNavigation( { waitUntil: 'load' } );

		await expect( page.title() ).resolves.toMatch( 'New title' );

		// Undo the changess
		await write( file, content );
		try {
			await unlink( to_create );
		} catch ( error ) {
		    // continue
		}

		done();

	});

	test('7. Current dir used when passing no parameters', async ( done ) => {

		const original_cwd = process.cwd();
		process.chdir( 'test/07/' )

		// If this test failed previously, lets guarantee that
		// everything is correct again before we begin
		const file    = 'index.html'
		const content = ( await read( file, 'utf8' ) ).replace( 'New title', 'Old title' );
		await write( file, content );

		await sleep( 130 )



		const server  = await liven();

		await page.goto( 'http://localhost:' + server.port );

		await expect( page.title() ).resolves.toMatch( 'Old title' );

		// Read and modify index.html to generate a refresh 
		await write( file, content.replace( 'Old title', 'New title' ) );

		// Wait the page to refresh
		await page.waitForNavigation( { waitUntil: 'load' } )

		await expect( page.title() ).resolves.toMatch( 'New title' );

		// Undo the changes
		await write( file, content );
		process.chdir( original_cwd )

		done();

	});

	test('8. Remap a diretory to another', async ( done ) => {

		const server  = await liven({
			dir: 'test/08',
			alias: {
				'images/': 'compressed/'
			}
		});

		await sleep( 130 )

		await page.goto( 'http://localhost:' + server.port + '/images/img.txt' );

		await expect( page ).toMatch( 'compressed image' )

		done();

	});

	test('9. Don\'t inject code on non-existent index (404)"', async done => {

		const server  = await liven( { dir: 'test/09' } );

		http.get( 'http://localhost:' + server.port, response => {

			expect( response.statusCode ).toBe( 404 );
			done();

		})

		done();

	});

	test('10. Memory INDEX files', async ( done ) => {

		const server = await liven( { dir: 'test/10' } );

		// index.html from the disk
		await page.goto( 'http://localhost:' + server.port );
		await expect( page ).toMatch( 'Hello world in the disk' );
		await sleep( 130 )

		// index.html from memory
		await server.memory( 'index.html', '<html><body> Hello world in memory! </body></html>' )
		await sleep( 130 )
		// await page.goto( 'http://localhost:' + server.port );
		await expect( page ).toMatch( 'Hello world in memory!' );


		// Updating index.html from memory
		await server.memory( 'index.html', '<html><body> Hello UPDATED world in memory! </body></html>' )
		await sleep( 130 )
		await expect( page ).toMatch( 'Hello UPDATED world in memory!' );


		// Clearing index.html, reading from the disk
		await server.clear( 'index.html' )
		await sleep( 130 )
		await expect( page ).toMatch( 'Hello world in the disk' );

		done();

	});

	test('11. Memory NON-INDEX files', async ( done ) => {

		const server = await liven( { dir: 'test/11' } );

		// index.html from the disk
		await page.goto( 'http://localhost:' + server.port + '/other.html' );
		await sleep( 130 )
		await expect( page ).toMatch( 'Hello other in the disk' );

		// other.html
		await server.memory( 'other.html', '<html><body> Hello other in memory! </body><script type="text/javascript">(new WebSocket("ws://"+(location.host))).onmessage=function( args ){ location.reload( true ) };</script></html>' )
		await sleep( 130 )
		await expect( page ).toMatch( 'Hello other in memory!' );

		// Updating other.html
		await server.memory( 'other.html', '<html><body> Hello other UPDATED in memory! </body><script type="text/javascript">(new WebSocket("ws://"+(location.host))).onmessage=function( args ){ location.reload( true ) };</script></html>' )
		await sleep( 130 )
		await expect( page ).toMatch( 'Hello other UPDATED in memory!' );

		// Clearing other.html, reading from the disk
		await server.clear( 'other.html' )
		await sleep( 130 )
		await expect( page ).toMatch( 'Hello other in the disk' );

		done();

	});


});

src/index.js
import CheapWatch 			from 'cheap-watch';
import { createServer } 	from 'http';
import getport				from 'get-port';
import { join, resolve }	from 'path';
import * as pulsa			from '@alumna/pulsa';
import WebSocket			from 'ws';

import * as fs 				from 'fs';

class Liven {

	constructor ( options ) {

		this.options        = options
		this.options.script = '<script>(new WebSocket("ws://"+(location.host))).onmessage=' + this.options.script + '</script>';
		this.index 			= join( this.options.dir, 'index.html' );

		const pulsa_options = {
			dir:   options.dir,
			spa:   options.spa,
			alias: options.alias
		}

		this.server = createServer( pulsa.serve( pulsa_options ) )
	}

	async start () {

		// Inject the refresh function before serving the content
		this.inject();

		// Get an available port to run
		this.port = await getport( { port: this.options.port || getport.makeRange( 3000, 3100 ) } );

		// Start servers
		this.server.listen( this.port )

		// Bind the socket server on the http one
		this.ws           = new WebSocket.Server( { server: this.server } );
		this.connections  = []

		// Add every new connection to the pool
		this.ws.on( 'connection', connection => this.connections.push( connection ) );

		// Asynchronously start the watcher
		return this.watch()

	}

	async watch () {

		// Create a Cheap-watch instance 
		this.cheapwatch = new CheapWatch( { dir: this.options.dir } );

		// Initialize the watching process
		await this.cheapwatch.init();

		this.cheapwatch.on( '+', args => this.check_event( args, true ) );
		this.cheapwatch.on( '-', args => this.check_event( args, false ) );

		return true;

	}

	async check_event ( args, add_or_update ) {

		if ( join( this.options.dir, args.path ) == this.index )
			this.inject();
		else
			pulsa.clear( join( this.options.dir, args.path ) )

		const data = { path: args.path, isDir: args.stats.isDirectory(), isFile: args.stats.isFile(), add_or_update, isNew: args.isNew ? true : false }

		this.check_on_event( data )

	}

	async memory ( path, content, clear = false ) {

		path = resolve( this.options.dir, path )

		if ( path == this.index )
			this.inject( content );
		else
			clear ? pulsa.clear( path ) : pulsa.memory( path, content )

		const data = {
			path: path.slice( this.options.dir ),
			isDir: false,
			isFile: true,
			add_or_update: !clear,
			isNew: !clear
		}

		return this.check_on_event( data )

	}

	async clear ( path ) {

		return this.memory( path, null, true )

	}

	async check_on_event ( data ) {

		if ( typeof this.options.on_event === 'function' && !( await this.options.on_event( data ) ) )
			return;
		
		return this.refresh( data );

	}

	// This function *is necessary* to inject the refresh function
	// that will receive the signal from socket
	inject ( content ) {

		if ( content || fs.existsSync( this.index ) ) {

			content = content || fs.readFileSync( this.index , 'utf8' );
			content = content.replace( '<body>', '<body>' + this.options.script )

			pulsa.memory( this.index, content )

		}

	}

	refresh ( data ) {

		for ( let i = this.connections.length - 1; i >= 0; i-- ) {

			if ( this.connections[i].readyState === WebSocket.OPEN )
				this.connections[i].send( JSON.stringify( data ) );

		}

	}

}

export default async function ( options = {} ) {

	const minimal 	= {
		script: 'function( args ){ location.reload( true ) };',
		spa: 	false,
		alias: 	null
	}

	Object.assign( minimal, options );

	minimal.dir 	= resolve( minimal.dir || '.' );
	const instance 	= new Liven( minimal )

	await instance.start();

	return instance;

};